packages/medusa/src/api/store/products/helpers.ts

```diff
@@ -8,3 +14,117 @@ export const refetchProduct = async (
 ) => {
   return await refetchEntity("product", idOrFilter, scope, fields)
 }
+
+type VariantInventoryType = {
+  variant_id: string
+  variant: { manage_inventory: boolean }
+  required_quantity: number
+  inventory: InventoryItemDTO
+}
+
+export const wrapVariantsWithInventoryQuantity = async (
+  req: MedusaRequest,
+  variants: {
+    id: string
+    inventory_quantity?: number
+    manage_inventory?: boolean
+  }[]
+) => {
+  const remoteQuery = req.scope.resolve(ContainerRegistrationKeys.REMOTE_QUERY)
+  const variantIds = variants.map((variant) => variant.id).flat(1)
+
+  if (!variantIds.length) {
+    return
+  }
+
+  if (!req.context?.stock_location_id?.length) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Stock locations are required to compute inventory`
+    )
+  }
+
+  const linkQuery = remoteQueryObjectFromString({
+    entryPoint: "product_variant_inventory_item",
+    variables: {
+      filters: { variant_id: variantIds },
+      inventory: {
+        filters: {
+          location_levels: {
+            location_id: req.context?.stock_location_id || [],
+          },
+        },
+      },
+    },
+    fields: [
+      "variant_id",
+      "variant.manage_inventory",
+      "required_quantity",
+      "inventory.*",
+      "inventory.location_levels.*",
+    ],
+  })
+
+  const links: VariantInventoryType[] = await remoteQuery(linkQuery)
+  const variantInventoriesMap = new Map<string, VariantInventoryType[]>()
+
+  links.forEach((link) => {
+    const array: VariantInventoryType[] =
+      variantInventoriesMap.get(link.variant_id) || []
+
+    array.push(link)
+
+    variantInventoriesMap.set(link.variant_id, array)
+  })
+
+  for (const variant of variants || []) {
```

Comment:
@riqwan maybe from here until the end of the loop can be extracted into a function, WDYT? Anyway I leave it up to you, it isn't a big deal TBH